// @databricks/sdk-experimental/examples/workspace/catalog/create-volume.ts
#!/usr/bin/env -S npx ts-node -r tsconfig-paths/register
/**
 * Example: Create a Unity Catalog volume
 *
 * This example demonstrates:
 * - Creating a managed volume in Unity Catalog
 * - Setting volume properties
 * - Listing volumes in a schema
 * - Cleaning up resources
 *
 * Prerequisites:
 * - DATABRICKS_HOST environment variable
 * - Valid authentication (PAT, OAuth M2M, or other)
 * - Unity Catalog enabled workspace
 * - An existing catalog and schema (or use TEST_CATALOG and TEST_SCHEMA env vars)
 *
 * Usage:
 *   export TEST_CATALOG="main"
 *   export TEST_SCHEMA="default"
 *   npx ts-node examples/workspace/catalog/create-volume.ts
 */

import {WorkspaceClient} from "@databricks/sdk-experimental";

async function main() {
    const client = new WorkspaceClient({});

    const catalogName = process.env.TEST_CATALOG || "main";
    const schemaName = process.env.TEST_SCHEMA || "default";
    const volumeName = `sdk_volume_${Date.now()}`;

    console.log(`Creating volume: ${catalogName}.${schemaName}.${volumeName}`);

    try {
        // Create a managed volume
        const volume = await client.volumes.create({
            catalog_name: catalogName,
            schema_name: schemaName,
            name: volumeName,
            volume_type: "MANAGED",
            comment: "Example volume created by SDK",
        });

        console.log(`✓ Volume created successfully`);
        console.log(`  Volume Name: ${volume.name}`);
        console.log(`  Full Name: ${volume.full_name}`);
        console.log(`  Volume Type: ${volume.volume_type}`);
        console.log(`  Storage Location: ${volume.storage_location || "N/A"}`);
        console.log(`  Owner: ${volume.owner}`);

        // Get volume details
        const volumeDetails = await client.volumes.read({
            name: volume.full_name!,
        });

        console.log(`\nVolume Details:`);
        console.log(`  Volume ID: ${volumeDetails.volume_id}`);
        console.log(
            `  Created At: ${volumeDetails.created_at ? new Date(volumeDetails.created_at).toISOString() : "N/A"}`
        );
        console.log(
            `  Updated At: ${volumeDetails.updated_at ? new Date(volumeDetails.updated_at).toISOString() : "N/A"}`
        );

        // List all volumes in the schema
        console.log(`\nListing all volumes in ${catalogName}.${schemaName}:`);
        const volumes = [];

        for await (const vol of client.volumes.list({
            catalog_name: catalogName,
            schema_name: schemaName,
        })) {
            volumes.push({
                "Name": vol.name,
                "Type": vol.volume_type,
                "Full Name": vol.full_name,
            });
        }

        if (volumes.length > 0) {
            console.table(volumes);
        } else {
            console.log("  No volumes found");
        }

        // Cleanup: Delete the volume
        console.log(`\nCleaning up: deleting volume ${volume.full_name}`);
        await client.volumes.delete({
            name: volume.full_name!,
        });
        console.log(`✓ Volume deleted`);
    } catch (error) {
        console.error(`✗ Error:`, error);
        process.exit(1);
    }
}

main();
