// @databricks/sdk-experimental/examples/workspace/catalog/manage-permissions.ts
#!/usr/bin/env -S npx ts-node -r tsconfig-paths/register
/**
 * Example: Manage Unity Catalog permissions
 *
 * This example demonstrates:
 * - Granting permissions on catalog objects
 * - Listing effective permissions
 * - Updating grants
 * - Revoking permissions
 *
 * Prerequisites:
 * - DATABRICKS_HOST environment variable
 * - Valid authentication (PAT, OAuth M2M, or other)
 * - Unity Catalog enabled workspace
 * - Admin permissions to manage grants
 * - An existing catalog and schema (or use TEST_CATALOG and TEST_SCHEMA env vars)
 *
 * Usage:
 *   export TEST_CATALOG="main"
 *   export TEST_SCHEMA="default"
 *   npx ts-node examples/workspace/catalog/manage-permissions.ts
 */

import {WorkspaceClient} from "@databricks/sdk-experimental";

async function main() {
    const client = new WorkspaceClient({});

    const catalogName = process.env.TEST_CATALOG || "main";
    const schemaName = process.env.TEST_SCHEMA || "default";
    const schemaFullName = `${catalogName}.${schemaName}`;

    console.log(`Managing permissions for schema: ${schemaFullName}\n`);

    try {
        // Get current user to use as principal
        const currentUser = await client.currentUser.me();
        const principal = currentUser.userName;

        console.log(`Current user: ${principal}`);

        // Get effective permissions on the schema
        console.log(`\nGetting effective permissions on ${schemaFullName}...`);

        const effectivePermissions = await client.grants.getEffective({
            securable_type: "SCHEMA",
            full_name: schemaFullName,
            principal: principal,
        });

        console.log(`✓ Effective permissions retrieved`);
        if (effectivePermissions.privilege_assignments) {
            for (const assignment of effectivePermissions.privilege_assignments) {
                console.log(`  Principal: ${assignment.principal}`);
                if (assignment.privileges) {
                    console.log(
                        `  Privileges: ${assignment.privileges.join(", ")}`
                    );
                }
            }
        }

        // List all grants on the schema
        console.log(`\nListing all grants on ${schemaFullName}...`);

        const grants = await client.grants.get({
            securable_type: "SCHEMA",
            full_name: schemaFullName,
        });

        console.log(`✓ Grants retrieved`);
        if (grants.privilege_assignments) {
            console.log(
                `  Total grant assignments: ${grants.privilege_assignments.length}`
            );

            for (const assignment of grants.privilege_assignments.slice(0, 5)) {
                console.log(`\n  Principal: ${assignment.principal}`);
                if (assignment.privileges) {
                    console.log(
                        `  Privileges: ${assignment.privileges.join(", ")}`
                    );
                }
            }

            if (grants.privilege_assignments.length > 5) {
                console.log(
                    `\n  ... (showing first 5 of ${grants.privilege_assignments.length})`
                );
            }
        } else {
            console.log(`  No grants found`);
        }

        // Example: Grant and revoke (commented out to avoid modifying permissions)
        // Uncomment to test granting/revoking permissions

        /*
        console.log(`\nGranting SELECT privilege to ${principal}...`);

        await client.grants.update({
            securable_type: "SCHEMA",
            full_name: schemaFullName,
            changes: [
                {
                    principal: principal,
                    add: ["SELECT"],
                },
            ],
        });

        console.log(`✓ SELECT privilege granted`);

        // Revoke the privilege
        console.log(`\nRevoking SELECT privilege from ${principal}...`);

        await client.grants.update({
            securable_type: "SCHEMA",
            full_name: schemaFullName,
            changes: [
                {
                    principal: principal,
                    remove: ["SELECT"],
                },
            ],
        });

        console.log(`✓ SELECT privilege revoked`);
        */

        console.log(`\n✓ Permission management example complete`);
        console.log(
            `  Note: Grant/revoke operations are commented out to avoid modifying permissions`
        );
    } catch (error) {
        console.error(`✗ Error:`, error);
        process.exit(1);
    }
}

main();
