// @databricks/sdk-experimental/examples/workspace/jobs/run-now.ts
#!/usr/bin/env -S npx ts-node -r tsconfig-paths/register
/**
 * Example: Trigger an existing job with parameters
 *
 * This example demonstrates:
 * - Triggering an existing job definition
 * - Passing runtime parameters to the job
 * - Monitoring the run execution
 * - Getting the run results
 *
 * Prerequisites:
 * - DATABRICKS_HOST environment variable
 * - Valid authentication (PAT, OAuth M2M, or other)
 * - An existing job ID (set TEST_JOB_ID environment variable)
 *
 * Usage:
 *   export TEST_JOB_ID="your-job-id"
 *   npx ts-node examples/workspace/jobs/run-now.ts
 */

import {WorkspaceClient, Time, TimeUnits} from "@databricks/sdk-experimental";

async function main() {
    const client = new WorkspaceClient({});

    const jobId = process.env.TEST_JOB_ID;

    if (!jobId) {
        console.error("✗ Error: TEST_JOB_ID environment variable not set");
        console.error("  Please set it to an existing job ID.");
        process.exit(1);
    }

    console.log(`Triggering job: ${jobId}`);

    try {
        // Get job details
        const job = await client.jobs.get({job_id: parseInt(jobId)});
        console.log(`Job name: ${job.settings?.name}`);

        // Trigger the job with parameters
        const waiter = await client.jobs.runNow({
            job_id: parseInt(jobId),
            // Optional: pass parameters to the job
            notebook_params: {
                environment: "development",
                run_date: new Date().toISOString().split("T")[0],
            },
        });

        console.log(`✓ Job triggered successfully`);
        console.log(`  Run ID: ${waiter.run_id}`);

        // Wait for completion
        console.log(`\nWaiting for job completion...`);

        const result = await waiter.wait({
            timeout: new Time(20, TimeUnits.minutes),
            onProgress: async (run) => {
                console.log(`  State: ${run.state?.life_cycle_state}`);
            },
        });

        console.log(`\n✓ Job run completed!`);
        console.log(`  Result State: ${result.state?.result_state}`);
        console.log(
            `  Run URL: ${client.config.host}#job/${jobId}/run/${result.run_id}`
        );

        if (result.state?.result_state === "SUCCESS") {
            console.log(`  Status: ✓ Job succeeded`);
        } else {
            console.log(
                `  Status: ✗ Job ${result.state?.result_state?.toLowerCase()}`
            );
        }
    } catch (error) {
        console.error(`✗ Error:`, error);
        process.exit(1);
    }
}

main();
