// @databricks/sdk-experimental/examples/workspace/files/dbfs-download.ts
#!/usr/bin/env -S npx ts-node -r tsconfig-paths/register
/**
 * Example: Download a file from DBFS
 *
 * This example demonstrates:
 * - Reading file content from DBFS
 * - Decoding base64 content
 * - Reading files in chunks
 * - Handling different file types
 * - Checking file existence
 *
 * Prerequisites:
 * - DATABRICKS_HOST environment variable
 * - Valid authentication (PAT, OAuth M2M, or other)
 * - An existing DBFS file path (or the script will create a temporary one)
 *
 * Usage:
 *   export TEST_DBFS_FILE="/tmp/your-file.txt"
 *   npx ts-node examples/workspace/files/dbfs-download.ts
 */

import {WorkspaceClient} from "@databricks/sdk-experimental";

async function main() {
    const client = new WorkspaceClient({});

    let dbfsPath = process.env.TEST_DBFS_FILE;
    let createdFile = false;

    // If no file path provided, create a temporary one
    if (!dbfsPath) {
        dbfsPath = `/tmp/sdk-download-example-${Date.now()}.txt`;
        createdFile = true;

        console.log(`Creating temporary file: ${dbfsPath}`);

        const content = `This is a test file for download demonstration.
Created at: ${new Date().toISOString()}

Lorem ipsum dolor sit amet, consectetur adipiscing elit.
This file contains multiple lines of text to demonstrate
downloading and reading file content from DBFS.
`;

        await client.dbfs.put({
            path: dbfsPath,
            contents: Buffer.from(content).toString("base64"),
            overwrite: true,
        });

        console.log(`✓ Temporary file created\n`);
    }

    console.log(`Downloading file from DBFS: ${dbfsPath}\n`);

    try {
        // Check if file exists
        try {
            const status = await client.dbfs.getStatus({
                path: dbfsPath,
            });

            console.log(`File Status:`);
            console.log(`  Path: ${status.path}`);
            console.log(`  Size: ${status.file_size} bytes`);
            console.log(`  Is Directory: ${status.is_dir}`);
            console.log(
                `  Modified: ${status.modification_time ? new Date(status.modification_time).toISOString() : "N/A"}\n`
            );

            if (status.is_dir) {
                console.error(`✗ Error: Path is a directory, not a file`);
                process.exit(1);
            }
        } catch (error) {
            console.error(`✗ Error: File not found at ${dbfsPath}`);
            process.exit(1);
        }

        // Read file content
        console.log(`Reading file content...`);

        const fileContent = await client.dbfs.read({
            path: dbfsPath,
        });

        if (fileContent.data) {
            // Decode base64 content
            const decodedContent = Buffer.from(
                fileContent.data,
                "base64"
            ).toString("utf-8");

            console.log(`✓ File downloaded successfully`);
            console.log(`  Bytes Read: ${fileContent.bytes_read}`);
            console.log(
                `  Content length: ${decodedContent.length} characters\n`
            );

            console.log(`File Content:`);
            console.log(`${"=".repeat(60)}`);
            console.log(decodedContent);
            console.log(`${"=".repeat(60)}\n`);
        }

        // Example: Read file in chunks (useful for large files)
        console.log(`--- Reading file in chunks (offset/length) ---\n`);

        // Read first 50 bytes
        const chunk = await client.dbfs.read({
            path: dbfsPath,
            offset: 0,
            length: 50,
        });

        if (chunk.data) {
            const chunkContent = Buffer.from(chunk.data, "base64").toString(
                "utf-8"
            );

            console.log(`First 50 bytes:`);
            console.log(`  "${chunkContent}"`);
            console.log(`  Bytes read: ${chunk.bytes_read}\n`);
        }

        // Example: Create and download a JSON file
        console.log(`--- JSON file example ---\n`);

        const jsonPath = `/tmp/sdk-data-${Date.now()}.json`;
        const jsonData = {
            name: "SDK Example",
            values: [10, 20, 30, 40, 50],
            timestamp: new Date().toISOString(),
            nested: {
                property1: "value1",
                property2: "value2",
            },
        };

        // Upload JSON
        await client.dbfs.put({
            path: jsonPath,
            contents: Buffer.from(JSON.stringify(jsonData, null, 2)).toString(
                "base64"
            ),
            overwrite: true,
        });

        console.log(`Created JSON file: ${jsonPath}`);

        // Download and parse JSON
        const jsonContent = await client.dbfs.read({
            path: jsonPath,
        });

        if (jsonContent.data) {
            const jsonString = Buffer.from(jsonContent.data, "base64").toString(
                "utf-8"
            );
            const parsed = JSON.parse(jsonString);

            console.log(`✓ JSON file downloaded and parsed:`);
            console.log(JSON.stringify(parsed, null, 2));
        }

        // Cleanup
        console.log(`\nCleaning up...`);

        if (createdFile) {
            await client.dbfs.delete({
                path: dbfsPath,
                recursive: false,
            });
            console.log(`✓ Deleted temporary file: ${dbfsPath}`);
        }

        await client.dbfs.delete({
            path: jsonPath,
            recursive: false,
        });
        console.log(`✓ Deleted JSON file: ${jsonPath}`);

        console.log(`\n✓ Download examples complete`);
    } catch (error) {
        console.error(`✗ Error:`, error);

        // Cleanup on error
        if (createdFile && dbfsPath) {
            try {
                await client.dbfs.delete({
                    path: dbfsPath,
                    recursive: false,
                });
            } catch (e) {
                // Ignore cleanup errors
            }
        }

        process.exit(1);
    }
}

main();
