// @databricks/sdk-experimental/examples/workspace/notebooks/export-notebook.ts
#!/usr/bin/env -S npx ts-node -r tsconfig-paths/register
/**
 * Example: Export a notebook from Databricks workspace
 *
 * This example demonstrates:
 * - Exporting a notebook in different formats (SOURCE, HTML, JUPYTER, DBC)
 * - Decoding base64 content
 * - Saving exported content
 * - Listing workspace objects
 *
 * Prerequisites:
 * - DATABRICKS_HOST environment variable
 * - Valid authentication (PAT, OAuth M2M, or other)
 * - An existing notebook path (set TEST_NOTEBOOK_PATH environment variable,
 *   or the script will create a temporary one)
 *
 * Usage:
 *   export TEST_NOTEBOOK_PATH="/Shared/your-notebook"
 *   npx ts-node examples/workspace/notebooks/export-notebook.ts
 */

import {WorkspaceClient} from "@databricks/sdk-experimental";
import * as fs from "fs";
import * as path from "path";

async function main() {
    const client = new WorkspaceClient({});

    let notebookPath = process.env.TEST_NOTEBOOK_PATH;
    let createdNotebook = false;

    // If no notebook path provided, create a temporary one
    if (!notebookPath) {
        notebookPath = `/Shared/sdk-export-example-${Date.now()}`;
        createdNotebook = true;

        console.log(`Creating temporary notebook: ${notebookPath}`);

        const content = `# Databricks notebook source
print("This is an example notebook for export")

# COMMAND ----------

display(spark.range(5))
`;

        await client.workspace.import({
            path: notebookPath,
            format: "SOURCE",
            language: "PYTHON",
            content: Buffer.from(content).toString("base64"),
            overwrite: true,
        });

        console.log(`✓ Temporary notebook created\n`);
    }

    console.log(`Exporting notebook: ${notebookPath}\n`);

    try {
        // Export in SOURCE format
        console.log(`1. Exporting as SOURCE format...`);

        const sourceExport = await client.workspace.export({
            path: notebookPath,
            format: "SOURCE",
        });

        if (sourceExport.content) {
            const sourceContent = Buffer.from(
                sourceExport.content,
                "base64"
            ).toString("utf-8");

            console.log(`✓ Exported SOURCE format`);
            console.log(`  Content length: ${sourceContent.length} characters`);
            console.log(
                `  First 200 chars:\n${sourceContent.substring(0, 200)}...\n`
            );
        }

        // Export in HTML format
        console.log(`2. Exporting as HTML format...`);

        const htmlExport = await client.workspace.export({
            path: notebookPath,
            format: "HTML",
        });

        if (htmlExport.content) {
            const htmlContent = Buffer.from(
                htmlExport.content,
                "base64"
            ).toString("utf-8");

            console.log(`✓ Exported HTML format`);
            console.log(`  Content length: ${htmlContent.length} characters`);
            console.log(`  Contains HTML: ${htmlContent.includes("<html>")}\n`);

            // Optionally save to file
            // const outputPath = path.join(process.cwd(), "exported-notebook.html");
            // fs.writeFileSync(outputPath, htmlContent);
            // console.log(`  Saved to: ${outputPath}\n`);
        }

        // Export in JUPYTER format
        console.log(`3. Exporting as JUPYTER format...`);

        const jupyterExport = await client.workspace.export({
            path: notebookPath,
            format: "JUPYTER",
        });

        if (jupyterExport.content) {
            const jupyterContent = Buffer.from(
                jupyterExport.content,
                "base64"
            ).toString("utf-8");

            console.log(`✓ Exported JUPYTER format`);
            console.log(
                `  Content length: ${jupyterContent.length} characters`
            );

            // Parse and show structure
            try {
                const notebook = JSON.parse(jupyterContent);
                console.log(
                    `  Notebook version: ${notebook.nbformat}.${notebook.nbformat_minor}`
                );
                console.log(
                    `  Number of cells: ${notebook.cells?.length || 0}\n`
                );
            } catch (e) {
                console.log(`  (Could not parse as JSON)\n`);
            }
        }

        // Export in DBC format (Databricks archive)
        console.log(`4. Exporting as DBC format...`);

        const dbcExport = await client.workspace.export({
            path: notebookPath,
            format: "DBC",
        });

        if (dbcExport.content) {
            console.log(`✓ Exported DBC format`);
            console.log(`  Content is binary (DBC archive)`);
            console.log(
                `  Base64 length: ${dbcExport.content.length} characters\n`
            );
        }

        // List workspace to show notebook
        console.log(`5. Listing workspace objects in /Shared...`);

        const objects: any[] = [];
        for await (const obj of client.workspace.list({
            path: "/Shared",
        })) {
            objects.push({
                Path: obj.path,
                Type: obj.object_type,
            });
            if (objects.length >= 5) break; // Limit to first 5
        }

        if (objects.length > 0) {
            console.table(objects);
        } else {
            console.log(`  No objects found`);
        }

        // Cleanup if we created a temporary notebook
        if (createdNotebook) {
            console.log(`\nCleaning up: deleting temporary notebook`);
            await client.workspace.delete({
                path: notebookPath!,
                recursive: false,
            });
            console.log(`✓ Temporary notebook deleted`);
        }

        console.log(`\n✓ Export examples complete`);
    } catch (error) {
        console.error(`✗ Error:`, error);

        // Cleanup on error
        if (createdNotebook && notebookPath) {
            try {
                await client.workspace.delete({
                    path: notebookPath,
                    recursive: false,
                });
            } catch (e) {
                // Ignore cleanup errors
            }
        }

        process.exit(1);
    }
}

main();
